homeassistant/components/dsmr/config_flow.py

```diff
@@ -1,15 +1,108 @@
 """Config flow for DSMR integration."""
+import asyncio
+from functools import partial
 import logging
 from typing import Any, Dict, Optional
 
-from homeassistant import config_entries
+from dsmr_parser import obis_references as obis_ref
+from dsmr_parser.clients.protocol import create_dsmr_reader, create_tcp_dsmr_reader
+import serial
+
+from homeassistant import config_entries, core, exceptions
 from homeassistant.const import CONF_HOST, CONF_PORT
 
-from .const import DOMAIN  # pylint:disable=unused-import
+from .const import (  # pylint:disable=unused-import
+    CONF_DSMR_VERSION,
+    CONF_SERIAL_ID,
+    CONF_SERIAL_ID_GAS,
+    DOMAIN,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 
+class DSMRConnection:
+    """Test the connection to DSMR and receive telegram to read serial ids."""
+
+    def __init__(self, host, port, dsmr_version):
+        """Initialize."""
+        self._host = host
+        self._port = port
+        self._dsmr_version = dsmr_version
+        self._telegram = {}
+
+    def equipment_identifier(self):
+        """Equipment identifier."""
+        if obis_ref.EQUIPMENT_IDENTIFIER in self._telegram:
+            dsmr_object = self._telegram[obis_ref.EQUIPMENT_IDENTIFIER]
+            return getattr(dsmr_object, "value", None)
+
+    def equipment_identifier_gas(self):
+        """Equipment identifier gas."""
+        if obis_ref.EQUIPMENT_IDENTIFIER_GAS in self._telegram:
+            dsmr_object = self._telegram[obis_ref.EQUIPMENT_IDENTIFIER_GAS]
+            return getattr(dsmr_object, "value", None)
+
+    async def validate_connect(self, hass: core.HomeAssistant) -> bool:
+        """Test if we can validate connection with the device."""
+
+        def update_telegram(telegram):
+            self._telegram = telegram
+
+            transport.close()
+
+        if self._host is None:
+            reader_factory = partial(
+                create_dsmr_reader,
+                self._port,
+                self._dsmr_version,
+                update_telegram,
+                loop=hass.loop,
+            )
+        else:
+            reader_factory = partial(
+                create_tcp_dsmr_reader,
+                self._host,
+                self._port,
+                self._dsmr_version,
+                update_telegram,
+                loop=hass.loop,
+            )
+
+        try:
```

Comment:
Would it be better to wrap this in a asyncio.wait_for? The dsmr meter reports every 10 seconds or faster. I've noticed that if I use the adress of another serial device not pushing data, it endlessly waits for data. Or is there another way to run a task with timeout?

Reply:
Use `async_timeout`. It's a core dependency.

https://pypi.org/project/async-timeout/

Reply:
Thanks! I've added the code, need to add a test for it and manually test it.